### PR TITLE
Sdi 64 paged mappings

### PR DESCRIPTION
**Pull request by lbennett-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/PageableWebFluxConfig.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import org.springframework.data.web.ReactivePageableHandlerMethodArgumentResolver
+import org.springframework.web.reactive.config.WebFluxConfigurer
+import org.springframework.web.reactive.result.method.annotation.ArgumentResolverConfigurer
+
+@Configuration
+class PageableWebFluxConfig : WebFluxConfigurer {
+  override fun configureArgumentResolvers(configurer: ArgumentResolverConfigurer) {
+    configurer.addCustomResolver(ReactivePageableHandlerMethodArgumentResolver())
+  }
+}
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/PageableWebFluxConfig.kt:13]: Required for the Pageable parameter to be resolved in the resource correctly

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/jpa/repository/VisitIdReactiveRepository.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.MappingType
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.VisitId
+
+@Repository
+interface VisitIdReactiveRepository : ReactiveSortingRepository<VisitId, Long> {
+  fun findAllByLabelAndMappingTypeOrderByLabelDesc(label: String, mappingType: MappingType, pageable: Pageable): Flux<VisitId>
+  fun countAllByLabelAndMappingType(label: String, mappingType: MappingType): Mono<Long>
+}
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/jpa/repository/VisitIdReactiveRepository.kt:15]: Had to be a separate repository to the co-routine version as the functions clashed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/MappingResource.kt [hunk 3]
```diff
@@ -164,4 +168,32 @@ class MappingResource(private val mappingService: MappingService) {
...
+    ]
+  )
+  suspend fun getVisitMappingsByMigrationId(
+    @PageableDefault pageRequest: Pageable,
+    @Schema(description = "Migration Id", example = "2020-03-24T12:00:00", required = true)
+    @PathVariable migrationId: String
+  ): Page<MappingDto> {
+    return mappingService.getVisitMappingsByMigrationId(pageRequest = pageRequest, migrationId = migrationId)
+      .awaitSingle()
+  }
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/MappingResource.kt:198]: awaitSingle for conversion from flux for the suspend function (can't have a non suspend function unless its a separate resource)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/MappingService.kt [hunk 4]
```diff
@@ -63,6 +69,28 @@ class MappingService(
...
   suspend fun deleteVisitMappings() = visitIdRepository.deleteAll()
+
+  fun getVisitMappingsByMigrationId(pageRequest: Pageable, migrationId: String): Mono<Page<MappingDto>> {
+    return visitIdReactiveRepository.findAllByLabelAndMappingTypeOrderByLabelDesc(
+      label = migrationId,
+      MappingType.MIGRATED,
+      pageRequest
+    ).collectList()
+      .zipWith(visitIdReactiveRepository.countAllByLabelAndMappingType(migrationId, mappingType = MappingType.MIGRATED))
+      .map { t ->
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/MappingService.kt:80]: separate count query required - int test checks the counts match the content in the response

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/MappingResource.kt (hunk 3 above)
Comment (lbennett-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/MappingResource.kt:198]: there might be an alternative to awaitSingle ...

Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/MappingResource.kt:198]: extremely minor, but might be slightly better for the awaitSingle to move down to the service

**Review by andymarke — APPROVED** [review]
LGTM

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**